### PR TITLE
Issue #423 Solved  -> Dynamically generating colors of all the lables…

### PR DESCRIPTION
**Pull request by aashu0148**
… in /zindex plugin are now maintain their legibility contrast and are easily visible

### CODE REVIEW

**Comment by aashu0148** [comment]
@googlebot I signed it!

**Comment by argyleink** [comment]
it is significantly improved! great work 👍 

there are still 1/15 or so that fail, any thoughts on why or if you could tweak it to be even more conservative? 

Screen Shot 2020-11-24 at 2 58 46 PM
Screen Shot 2020-11-24 at 2 58 52 PM

maybe switch to hsl to rotate the hue for randomization? 
or create a "safe palette" of like 50 colors and randomly pick one?

**Comment by aashu0148** [comment]
> it is significantly improved! great work 👍
> 
> there are still 1/15 or so that fail, any thoughts on why or if you could tweak it to be even more conservative?
> 
> Screen Shot 2020-11-24 at 2 58 46 PM
> Screen Shot 2020-11-24 at 2 58 52 PM
> 
> maybe switch to hsl to rotate the hue for randomization?
> or create a "safe palette" of like 50 colors and randomly pick one?

okay @argyleink  I'll modify the colors to be more darker.

**Comment by aashu0148** [comment]
so @argyleink should i close this pull request and make a new one.

**Comment by aashu0148** [comment]
hey @argyleink  please review it. Now I have made all the colors little bit darker.

**Comment by argyleink** [comment]
Gave it a nice look and review today! 

I was able to find labels quickly that didn't pass the contrast ratio scores.. I tried tweaking your numbers in the code here and there to see if a healthy value showed up, but it felt like it was ruining the value of the different colors, like they were merging towards middle ranges. 

I think we need to stop "randomly" making colors, hoping they pass. Here's 2 ideas forward: 
1. create ~25 colors we think provide enough differentiation and will contrast well with white, then pick randomly from that set.
1. import the tiny color library that visbug uses to modify and inspect colors. use this library to create random colors and test them for passing ratios. run this in a loop so it always returns a random, valid contrast passing label background color.

1 is simpler and static. 2 would be trickier but more dynamic. thoughts?

**Review by argyleink — CHANGES_REQUESTED** [review]
to close the ticket with confidence, we need to always create readable labels

**Comment by aashu0148** [comment]
> Gave it a nice look and review today!
> 
> I was able to find labels quickly that didn't pass the contrast ratio scores.. I tried tweaking your numbers in the code here and there to see if a healthy value showed up, but it felt like it was ruining the value of the different colors, like they were merging towards middle ranges.
> 
> I think we need to stop "randomly" making colors, hoping they pass. Here's 2 ideas forward:
> 
> 1. create ~25 colors we think provide enough differentiation and will contrast well with white, then pick randomly from that set.
> 2. import the tiny color library that visbug uses to modify and inspect colors. use this library to create random colors and test them for passing ratios. run this in a loop so it always returns a random, valid contrast passing label background color.
> 
> 1 is simpler and static. 2 would be trickier but more dynamic. thoughts?

okay @argyleink  i will check try both ways and update you with suitable one.

**Comment by aashu0148** [comment]
hey @argyleink where do i find tiny color library ??

**Comment by aashu0148** [comment]
> Gave it a nice look and review today!
> 
> I was able to find labels quickly that didn't pass the contrast ratio scores.. I tried tweaking your numbers in the code here and there to see if a healthy value showed up, but it felt like it was ruining the value of the different colors, like they were merging towards middle ranges.
> 
> I think we need to stop "randomly" making colors, hoping they pass. Here's 2 ideas forward:
> 
> 1. create ~25 colors we think provide enough differentiation and will contrast well with white, then pick randomly from that set.
> 2. import the tiny color library that visbug uses to modify and inspect colors. use this library to create random colors and test them for passing ratios. run this in a loop so it always returns a random, valid contrast passing label background color.
> 
> 1 is simpler and static. 2 would be trickier but more dynamic. thoughts?

hey @argyleink I have applied method 1 to it. Added a array of 25 colors and then randomly choosing a color from it.
Please check this :)

File: app/plugins/zindex.js [hunk 1]
```diff
@@ -4,14 +4,17 @@ export const commands = [
...
+    return Math.floor(min + (Math.random() * (max - min)))
+}
 export default function () {
     // Fun prior art https://gist.github.com/paulirish/211209
     Array.from(document.querySelectorAll('*'))
         .filter(el => el.computedStyleMap().get('z-index').value !== 'auto')
         .filter(el => el.nodeName !== 'VIS-BUG')
         .forEach(el => {
-            const color = `#${Math.floor(Math.random() * 16777215).toString(16).padStart(6,0)}`
+            const colors = ["#eb4034", "#68eb34", "#34a5eb", "#4334eb", "#b134eb", "#eb34a2", "#eb3446", "#8f2e2b", , "#8f692b", "#8a8f2b", "#358f2b", "#2b8f82", "#2b678f", "#2b2b8f", "#8f2b8f", "#8f2b55", "#1eff00", "#ff9d00", "#ff0000", "#00ff6a", "#0026ff", "#bb00ff", "#ff00d9", "#ff0073", "#ff0073", "#137878"];
```
Comment (argyleink) [inline app/plugins/zindex.js:16]:
```suggestion
            const colors = ["#eb4034", "#30850f", "#116da7", "#4334eb", "#b134eb", "#df168e", "#e8172c", "#8f2e2b", "#8f692b", "#8a8f2b", "#358f2b", "#2b8f82", "#2b678f", "#2b2b8f", "#8f2b8f", "#8f2b55", "#1eff00", "#a86800", "#ff0000", "#008035", "#0026ff", "#bb00ff", "#d600b6", "#e60067", "#137878"];
```

**Review by argyleink — APPROVED** [review]
There were still some colors that didn't pass the contrast ratios, I've updated them 👍🏻  That puts this PR into a ready state!

**Comment by argyleink** [comment]
fixes #423

**Comment by argyleink** [comment]
nice work, thank you @aashu0148 💯

**Comment by aashu0148** [comment]
> nice work, thank you @aashu0148 💯 

Please tell me if anything more to do.... I will love to contribute to visbug😊

**Comment by argyleink** [comment]
Here's a list of [good first issues](https://github.com/GoogleChromeLabs/ProjectVisBug/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue+%F0%9F%91%8D%22) if you're looking for low hanging fruit!

There's also this one which could be nice and swift to work on https://github.com/GoogleChromeLabs/ProjectVisBug/issues/389, which the code is pretty straight forward but testing it may be difficult if you dont have windows/linux readily available.

There's also room for writing more tests. So if you're interested in unit or e2e tests, you could write some of those for practice/learning/helping. 

🙂

**Pull request merged**